### PR TITLE
Add fixed header and write skipping to multiplexer

### LINKED ISSUE #35647: SSH listener violates spec by waiting for client data
To support signed client source IP address propagation, in Teleport v14 the listener for the SSH service (when not running in tunnel mode) was wrapped in our generic protocol multiplexer to be able to parse the signed PROXY protocol v2 header coming from the Teleport Proxy. This, however, leads to a violation of the SSH spec, as the generic protocol multiplexer always waits for at least 8 bytes from the client (checking for a `SSH-` prefix, to confirm that it's indeed a SSH connection), even though [RFC 4253 § 4.2](https://datatracker.ietf.org/doc/html/rfc4253#section-4.2) says:

>When the connection has been established, both sides MUST send an identification string. This identification string MUST be `SSH-protoversion-softwareversion SP comments CR LF` [...] The server MAY send other lines of data before sending the version string.

The RFC isn't totally explicit about it, but as there's no indication of which side should send data first nor any implicit sequencing in the protocol (and because it makes perfect sense from a performance point of view) the most reasonable interpretation is that both client and server should be sending the version identification string at the same time. This is supported by the behavior of OpenSSH's `sshd` and the default behavior of `golang.org/x/crypto/ssh.NewServerConn`: send the version string and then read the version from the client - socket buffering turns this into effectively a simultaneous send and receive.

This is not a problem with most SSH clients, as they also typically employ the same strategy and will thus work just fine with Teleport v14. However, two notable examples that violate the spec (just like Teleport v14's SSH server) by waiting for the server to send _anything_ before sending their version string are OpenSSH's very own `ssh-keyscan` and [`pdsh`](https://github.com/chaos/pdsh).

As the only possible underlying listener for the protocol multiplexer on the SSH service is the SSH server and the PROXY protocol doesn't require any acknowledgment from the server, we should send the SSH version string at connection unconditionally first, and only then potentially parse the PROXY protocol header (and close the connection if the signature check fails, if necessary). In addition to bringing the server back into spec, this should also improve handshake times by ~half a roundtrip or so, as the client won't need to wait for the server to reply to the version identification before continuing the handshake.

### PR DESCRIPTION
**Pull request by espadolini**
Add the ability to unconditionally send some fixed header to the client before protocol detection in `multiplexer.Mux`, together with the ability to skip the same amount of data sent from the application side; this lets us send a SSH version identification header before we run PROXY header detection, where appropriate.

Fixes #35647.

Changelog: Restored direct dial SSH server compatibility with certain SSH tools such as `ssh-keyscan` (#35647)

### CODE REVIEW

File: lib/multiplexer/multiplexer_test.go [hunk 1]
```diff
@@ -1395,3 +1395,53 @@ func TestIsDifferentTCPVersion(t *testing.T) {
 			fmt.Sprintf("Unexpected result for %q, %q", tt.addr1, tt.addr2))
 	}
 }
+
+func TestFixedHeader(t *testing.T) {
+	t.Parallel()
+	require := require.New(t)
```
Comment (AntonAM) [inline lib/multiplexer/multiplexer_test.go:1401]: TIL

File: lib/multiplexer/wrappers.go [hunk 2]
```diff
@@ -37,6 +37,7 @@ type Conn struct {
 	protocol  Protocol
 	proxyLine *ProxyLine
 	reader    *bufio.Reader
+	writeSkip int
```
Comment (AntonAM) [inline lib/multiplexer/wrappers.go:40]: Do you think just skipping amount of bytes without controlling those bytes match prefix is enough?
We most probably should be fine, but if something goes wrong there's possibility for some cryptic bugs 😅

**Review by AntonAM — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Review by smallinsky — APPROVED** [review]

File: lib/multiplexer/wrappers.go [hunk 3]
```diff
@@ -57,6 +58,28 @@ func (c *Conn) Read(p []byte) (int, error) {
...
+// Write implements [io.Writer] and [net.Conn].
+func (c *Conn) Write(p []byte) (int, error) {
+	if c.writeSkip < 1 {
+		return c.Conn.Write(p)
+	}
+
+	if len(p) <= c.writeSkip {
+		if _, err := c.Conn.Write(nil); err != nil {
+			return 0, trace.Wrap(err)
+		}
```
[the remark this answers is not in the document]
Comment (espadolini) [reply lib/multiplexer/wrappers.go:72]: Added this:
```go
		// we do a zero-length write here to check that the connection is still
		// open and that we're not past a write deadline
```

File: lib/multiplexer/wrappers.go (hunk 2 above)
Comment (espadolini) [reply lib/multiplexer/wrappers.go]: You're right, I added a check for that.

**Review by AntonAM — APPROVED** [review]

**Review by rosstimothy — APPROVED** [review]

**Pull request merged**